### DISCUSSION TITLE
Why tenant initializer is called mutiple times?

### DISCUSSION THREAD

**Discussion by xperiandri** [OrchardCMS/OrchardCore#14640] [Q&A]
I have 2 tenants:
* Default
* Sandbox

When I add initializer I see
___________________________________
info: OrchardCore.Environment.Shell.ShellHost[0]
      Start creation of shells
info: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating shell context for tenant 'Default'
dbug: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating described context for tenant 'Default'
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Composing blueprint
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Done composing blueprint
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:4651
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:4652
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\Andrii\Dev\Lula\risk-assessment\app\presentation\WebHost
Configuring 'self_signed_jwt' authentication strategy on Development environment.
**Initializer called for Default**
dbug: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating described context for tenant 'Default'
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Composing blueprint
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Done composing blueprint
Configuring 'self_signed_jwt' authentication strategy on Development environment.
**Initializer called for Default**
dbug: OrchardCore.Environment.Shell.ShellHost[0]
      Registering shell context for tenant 'Default'
dbug: OrchardCore.Environment.Shell.ShellHost[0]
      Registering shell context for tenant 'Sandbox'
info: OrchardCore.Environment.Shell.ShellHost[0]
      Done pre-creating and registering shells
dbug: OrchardCore.Environment.Shell.ShellHost[0]
      Creating shell context for tenant 'Default'
dbug: OrchardCore.Environment.Shell.ShellHost[0]
      Creating shell context for tenant 'Sandbox'
info: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating shell context for tenant 'Default'
info: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating shell context for tenant 'Sandbox'
dbug: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating described context for tenant 'Default'
dbug: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating described context for tenant 'Sandbox'
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Composing blueprint
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Done composing blueprint
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Composing blueprint
Configuring 'self_signed_jwt' authentication strategy on Development environment.
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Done composing blueprint
Configuring 'self_signed_jwt' authentication strategy on Development environment.
**Initializer called for Default**
**Initializer called for Sandbox**
dbug: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating described context for tenant 'Default'
dbug: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating described context for tenant 'Sandbox'
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Composing blueprint
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Composing blueprint
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Done composing blueprint
dbug: OrchardCore.Environment.Shell.Builders.CompositionStrategy[0]
      Done composing blueprint
**Initializer called for Sandbox**
**Initializer called for Default**
dbug: OrchardCore.Environment.Shell.ShellHost[0]
      Registering shell context for tenant 'Default'
dbug: OrchardCore.Environment.Shell.ShellHost[0]
      Registering shell context for tenant 'Sandbox'
________________________________________

Notice that the initializer for `Default` was called 4 times and for `Sandbox` 2 times.
Why does this happen and how to avoid that?

**Comment by xperiandri** [comment]
@jtkech any ideas?

**Comment by jtkech** [comment]
Before building a shell container we first create a light container to be able to query the database to retrieve the shell descriptor, but this 1st container is built with an empty descriptor without any feature so no startups are executed, and it doesn't activate the shell so no tenant event handlers are executed.

But yes if you register a container initializer `services.Initialize()` its delegate will be executed twice, we use this to initialize the YesSql store that we also need for the first light temporary container.

Then in the `DistributedShellHostedService` we also create an isolated container in the background (we only do that for the Default tenant), we use this to check if a possible distributed cache is registered and then use it (if the feature is enabled) to sync tenant states. So a container initializer is called 2 * 2 = 4 times

But recently we updated the `DistributedShellHostedService` so that the cloned shell is not built if the tenants distributed feature is not enabled, only the first container is built to check the shell descriptor.

So now, if the tenants distributed is not enabled, a container initializer is called 2 + 1 = 3 times, this while building 2 light temporary containers and only one full shell container.
